File: src/ChessChampions.rb
class Player
  attr_accessor :name, :age, :elo
  def initialize(name, age, elo)
    @name = name
    @age = age
    @elo = elo
  end

  def <=>(other)
    if self.age != other.age
      return other.age <=> self.age
    end
    self.elo <=> other.elo
  end
end

def find_champions(players)
  champions = []
  players.sort!
  best_player = Player.new("", 0, 0)
  players.each_with_index do |player, i|
    is_champion = !is_strictly_older_and_weaker(best_player, player)
    (i + 1).upto(players.size - 1) do |j|
      other = players[j]
      if !is_champion || is_older_and_weaker(player, other)
        unless has_same_age_and_elo(player, other)
          is_champion = false
        end
        best_player = set_best_player(other)
        break
      end
    end
    champions << player if is_champion
  end
  champions
end

def set_best_player(new_best_player)
  Player.new("BestPlayer", new_best_player.age, new_best_player.elo)
end

def has_same_age_and_elo(player, other)
  player.age == other.age && player.elo == other.elo
end

def is_older_and_weaker(player, other)
  player.age >= other.age && player.elo <= other.elo
end

def is_strictly_older_and_weaker(best_player, player)
  player.age > best_player.age && player.elo < best_player.elo
end

players = []
players << Player.new("player1", 30, 2800)
players << Player.new("player2", 25, 2700)
players << Player.new("player3", 32, 2900)
players << Player.new("player4", 28, 2600)
players << Player.new("player5", 17, 2700)
players << Player.new("player6", 17, 2700)
players << Player.new("player7", 16, 1000)
players << Player.new("player8", 16, 1200)
players << Player.new("player9", 46, 3000)
players << Player.new("player10", 47, 3400)
players << Player.new("player11", 46, 1)

champions = find_champions(players)
champions.each do |champion|
  puts "#{champion.name} #{champion.age} #{champion.elo}"
end

File: src/ChessChampions.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class ChessChampions {
    static class Player implements Comparable<Player> {
        String name;
        int age;
        int elo;

        public Player(String name, int age, int elo) {
            this.name = name;
            this.age = age;
            this.elo = elo;
        }

        @Override
        public int compareTo(Player other) {
            if (this.age != other.age) {
                return this.age - other.age;
            }
            return other.elo - this.elo;
        }
    }

    public static List<Player> findChampions(List<Player> players) {
        List<Player> champions = new ArrayList<>();
        Collections.sort(players);
        champions.add(players.get(0));
        for (int i = 1; i < players.size(); i++) {
            Player champion = champions.get(champions.size() - 1);
            Player player = players.get(i);
            if (isBetterThanTheChampion(champion, player)) {
                champions.add(player);
            }
        }
        return champions;
    }

    private static boolean isBetterThanTheChampion(Player champion, Player player) {
        return ((player.elo > champion.elo) || (player.age == champion.age && player.elo == champion.elo));
    }

    public static void main(String[] args) {
        List<Player> players = new ArrayList<>();
        players.add(new Player("player1", 30, 2800));
        players.add(new Player("player2", 25, 2700));
        players.add(new Player("player3", 32, 2900));
        players.add(new Player("player4", 28, 2600));
        players.add(new Player("player5", 17, 2700));
        players.add(new Player("player6", 17, 2700));
        players.add(new Player("player7", 16, 1000));
        players.add(new Player("player8", 16, 1200));
        players.add(new Player("player9", 46, 3000));
        players.add(new Player("player10", 47, 3400));
        players.add(new Player("player11", 46, 1));

        List<Player> champions = findChampions(players);
        for (Player champion : champions) {
            System.out.println(champion.name + " " + champion.age + " " + champion.elo);
        }
    }
}
